File: ruby/triangle/triangle.rb

class TriangleError < StandardError
end

class Triangle

  def initialize(a, b, c)
    @a, @b, @c = a, b, c
    ensure_valid_sides!
  end

  def kind
    case
    when ab? && bc?
      :equilateral
    when ab? || bc? || ac?
      :isosceles
    else
      :scalene
    end
  end

  private

  def ab?
    @a == @b
  end

  def bc?
    @b == @c
  end

  def ac?
    @a == @c
  end

  def ensure_valid_sides!
    invalid_side!(:a) unless @a > 0
    invalid_side!(:b) unless @b > 0
    invalid_side!(:c) unless @c > 0
    invalid_sides!    unless meets_inequality_constraint?
  end

  def invalid_side!(side)
    raise TriangleError, "Side #{side} must be > 0"
  end

  def invalid_sides!
    raise TriangleError, "Fails to meet inequality constraint"
  end

  def meets_inequality_constraint?
    ((@a-@b).abs < @c) && (@c < (@a+@b))
  end

end

File: elixir/triangle/triangle.exs
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: { :ok, kind } | { :error, String.t }

  def kind(a, b, c)
    when (a <= 0) or (b <= 0) or (c <= 0), do: {:error, "all side lengths must be positive"}

  def kind(a, b, c)
    when not (abs(a-b) < c and c < (a+b)), do: {:error, "side lengths violate triangle inequality"}

  def kind(_a, _a, _a), do: {:ok, :equilateral}

  def kind(_a, _a, _),  do: {:ok, :isosceles}
  def kind(_, _a, _a),  do: {:ok, :isosceles}
  def kind(_a, _, _a),  do: {:ok, :isosceles}

  def kind(_a, _b, _c), do: {:ok, :scalene}

end
